spec/http-gateway.did

```diff
@@ -0,0 +1,34 @@
+type HeaderField = record { text; text; };
+
+type HttpRequest = record {
+  method: text;
+  url: text;
+  headers: vec HeaderField;
+  body: blob;
+};
+
+type HttpResponse = record {
+  status_code: nat16;
+  headers: vec HeaderField;
+  body: blob;
+  upgrade : opt bool;
+  streaming_strategy: opt StreamingStrategy;
+};
+
+type StreamingCallbackHttpResponse = record {
+  body: blob;
+  token: opt T;
+};
+
+type StreamingStrategy = variant {
+  Callback: record {
+    callback: func (T) -> (opt StreamingCallbackHttpResponse) query;
```

Comment:
Unfortunately, there's a Candid bug which in practice will cause a runtime error if you define the callback like this.

See this [comment](https://forum.dfinity.org/t/cryptic-error-from-icx-proxy/6944/9?u=jzxchiang) by @chenyan-dfinity for details.

Perhaps the protocol can remain "pristine" but if users want code that actually works, they will need to change the type to:

`callback: func (?T) -> (opt StreamingCallbackHttpResponse);`

Reply:
Is that just due to bugs in some implementation that will be fixed in due time? Then I'd rather keep it as designed here, and leave temporary work-arounds out.

Reply:
It's only an implementation problem. There is a more recent workaround here by abusing serde's `ignored_any`: https://github.com/dfinity/agent-rs/blob/main/ic-utils/src/interfaces/http_request.rs#L231

I think the real problem here is that `T` is undefined, it's not a proper Candid file.

Reply:
Yeah, but that's too late to avoid now, right?

Reply:
I guess the solution is to catch up on the Candid side to add generics? In the Rust library, we kind of overcome this by deserializing the value into an untyped `IDLValue`.

In the meantime, we probably need to add a comment here to not confuse the reader.

Reply:
I wrote this in the adoc:

>  This type signature is a type signature _scheme_, and the canister can choose any type for the token type `T` used for streaming. The Gateway handles that type generically.

But yes, maybe better moved into the Candid file proper?

Reply:
Hmm, `reserved` isn’t quite right (that is a specific Candid type) But maybe
```
type Token = /* insert service-specific type here */
```
or something like that

Reply:
Refined, and added a stern warning